stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt

```diff
@@ -325,49 +289,49 @@ internal constructor(
      * @param tokenProvider A [TokenProvider] implementation.
      * @param listener Socket connection listener.
      */
-    private fun setUser(
+    private suspend fun setUser(
         user: User,
         tokenProvider: TokenProvider,
-        listener: InitConnectionListener? = null,
-    ) {
+    ): Result<ConnectionData> {
         val cacheableTokenProvider = CacheableTokenProvider(tokenProvider)
         if (tokenUtils.getUserId(cacheableTokenProvider.loadToken()) != user.id) {
             logger.logE("The user_id provided on the JWT token doesn't match with the current user you try to connect")
-            listener?.onError(
+            return Result.error(
                 ChatError(
                     "The user_id provided on the JWT token doesn't match with the current user you try to connect"
                 )
             )
-            return
         }
         val userState = userStateService.state
-        when {
+        return when {
             userState is UserState.UserSet &&
                 userState.user.id == user.id &&
                 socketStateService.state == SocketState.Idle -> {
+                logger.logV("[setUser] user is Set & socket.state is Idle")
                 userStateService.onUserUpdated(user)
                 tokenManager.setTokenProvider(cacheableTokenProvider)
-                connectionListener = listener
                 socketStateService.onConnectionRequested()
                 socket.connect(user)
                 initializationCoordinator.userConnected(user)
+                waitConnection.first()
```

Comment:
Can we guarantee that this value will always the non null?

Reply:
This contract comes from the deleted `InitConnectionListener `:
```kotlin
internal open class InitConnectionListener {

     open fun onSuccess(data: ConnectionData) {
     }

     open fun onError(error: ChatError) {
     }

     data class ConnectionData(val user: User, val connectionId: String)
}
```

So that's why `waitConnection` do not accept null values:
```kotlin
private val waitConnection = MutableSharedFlow<Result<ConnectionData>>()
```